### ISSUE TITLE
Recursive formatting

### ISSUE THREAD

**Issue by aheze** [aheze/Multiliner#3]
Currently when you select a chunk of code that contains parameters nested inside, formatting is weird:

<img width="700" alt="Memberwise init generated by Xcode" src="https://user-images.githubusercontent.com/49819455/176093790-0bd9203e-a305-4359-a40f-57c375bfb215.png">
 
This is because Multiliner splits parameters by commas `,` — this includes commas inside nested inits.

https://github.com/aheze/Multiliner/blob/main/Sources/MultilinerExtension/SourceEditorCommand.swift#L120-123

Possibly a split using Regex?

**Comment by calimarkus** [comment]
Maybe you could parse the code into an AST using [`apple/swift-syntax`](https://github.com/apple/swift-syntax) to understand the code better. The resulting type information could be helpful to make more informed decisions on how to rewrite the code. (Maybe you could also use this to identify the initial range?)

**Comment by aheze** [comment]
Worth a try. Not sure how the performance will be impacted though

**Comment by calimarkus** [comment]
It's designed for "performance-critical applications" - whatever that may mean: 

> SwiftSyntax is a set of Swift bindings for the [libSyntax](https://github.com/apple/swift/tree/main/lib/Syntax) library. [...] *Its API is designed for performance-critical applications.*

**Comment by aheze** [comment]
Right. Thanks for the suggestion, I'll do some experimenting this week!

**Comment by orchetect** [comment]
Yeah, because of assumptions the parsing makes, it can format things in an unexpected way so you have to be fairly specific about what you highlight sometimes. I find myself needing to hit <kbd>Ctrl</kbd>-`I` afterward which then indents the results correctly.

Adopting `swift-syntax` would be great - could help with the heavy lifting of handling recursive and complex formatting, and would evolve as the Swift language evolves too.

It would be hugely beneficial to be adding unit tests to ensure formatting continues to work as expected as the extension is updated over time, since there's so many edge cases and variations in code formatting.

**Comment by orchetect** [comment]
One major benefit to using a formal ASL is that you could format an entire file in one go. I often come across code I inherit or old code I wrote where parameters were all on single lines. It would be amazing to just <kbd>Cmd</kbd><kbd>A</kbd> and run the formatter.

**Comment by aheze** [comment]
@orchetect ah, so both SwiftFormats can kind of replace Multiliner. Nice investigating!

I guess if you want full file formatting it's better to go with one of the above. Multiliner was originally meant to be a quick shortcut to expand whatever line you're currently on. This makes it pretty fast actually (since it doesn't need to parse the whole file, it formats instantly even if your file is huge).

I've thought about whole file formatting, but the SwiftFormats can do this already and are great at it. Maybe it would be better to keep Multiliner lightweight and focus more on small shortcuts. I'll still look into AST though

**Comment by orchetect** [comment]
I don't think it's a replacement necessarily, after testing them out. It's also not exactly the same as Multiliner, you have to tweak the rules to get them to force it and it only works under certain circumstances. Whereas Multiliner does it every time. So I'd still probably use both depending on circumstances.

What I was thinking is it might be possible to learn from their source code how they're doing it.
